src/cpp/src/tokenizer/tokenizer.cpp

```diff
@@ -66,10 +66,55 @@ std::optional<std::string> remap_template(const std::string& chat_template) {
     return std::nullopt;
 }
 
-void parse_if_exists(const std::filesystem::path& path, std::string& value) {
-    if (std::filesystem::exists(path)) {
-        ov::genai::utils::read_json_param(nlohmann::json::parse(std::ifstream{path}), "chat_template", value);
+void parse_chat_template_from_file(const std::filesystem::path& path, std::string& value) {
+    if (!std::filesystem::exists(path)) {
+        return;
+    }
+    auto json_data = nlohmann::json::parse(std::ifstream{path});
+    if (!json_data.contains("chat_template")) {
+        return;
+    }
+    auto chat_template_field = json_data["chat_template"];
+
+    if (chat_template_field.is_string()) {
+        value = chat_template_field.get<std::string>();
+        return;
+    }
+    // Handle chat template format: [{ "name": "default", "template": "..." }]
+    // e.g. for CohereLabs/aya-23-8B & CohereLabs/c4ai-command-r-v01 models
+    if (chat_template_field.is_array()) {
+        for (const auto& item : chat_template_field) {
+            if (
+                item.is_object() && item.contains("name") && item["name"] == "default" &&
+                item.contains("template") && item["template"].is_string()
+            ) {
+                value = item["template"].get<std::string>();
+                return;
+            }
+        }
     }
+    OPENVINO_THROW("Unsupported chat_template format in file: ", path.string());
+}
+
+void parse_chat_template_from_tokenizer(std::shared_ptr<ov::Model> ov_tokenizer, std::string& value) {
+    if (!ov_tokenizer->has_rt_info("chat_template")) {
+        return;
+    }
+    ov::Any chat_template_value = ov_tokenizer->get_rt_info<ov::Any>("chat_template");
+    if (chat_template_value.is<std::string>()) {
+        value = chat_template_value.as<std::string>();
+        return;
+    }
+    // Handle rt_info chat template format: <chat_template><default value="..." /></chat_template>
+    if (chat_template_value.is<ov::AnyMap>()) {
+        ov::AnyMap chat_template_map = chat_template_value.as<ov::AnyMap>();
+        auto default_iter = chat_template_map.find("default");
+        if (default_iter != chat_template_map.end() && default_iter->second.is<std::string>()) {
+            value = default_iter->second.as<std::string>();
+            return;
+        }
+    }
+    OPENVINO_THROW("Unsupported type for 'chat_template' in tokenizer model: ", chat_template_value.type_info().name());
```

Comment:
@Wovchena Not sure if we need to throw here, maybe something like `std::cerr << [ WARNING ] ... << std::endl;` would be enough.
And the same question for `parse_chat_template_from_file()`.

Reply:
Warning message updated